drivers/i2c/i2c_numaker.c

```diff
@@ -0,0 +1,808 @@
+		break;
+		/* Slave Receive */
+	case S_RECE_DATA_ACK:  /* Slave Receive Data ACK */
+		data_byte = I2C_GET_DATA(i2c_base);
+		if (slave_callbacks->write_received(slave_config, data_byte) == 0) {
+			/* Write OK, ACK next data byte */
+			I2C_SET_CONTROL_REG(i2c_base, I2C_CTL0_SI_Msk | I2C_CTL0_AA_Msk);
+		} else {
+			/* Write FAILED, NACK next data byte */
+			I2C_SET_CONTROL_REG(i2c_base, I2C_CTL0_SI_Msk);
+		}
+		break;
+	case S_RECE_DATA_NACK:  /* Slave Receive Data NACK */
+		/* Go slave end */
+		data->slave_xfer.slave_addressed = false;
+		slave_callbacks->stop(slave_config);
+		I2C_SET_CONTROL_REG(i2c_base, I2C_CTL0_SI_Msk | I2C_CTL0_AA_Msk);
+		break;
+	case S_RECE_ADDR_ACK:  /* Slave Receive Address ACK */
+	case S_RECE_ARB_LOST:  /* Slave Receive Arbitration Lost */
+		data->slave_xfer.slave_addressed = true;
+		if (slave_callbacks->write_requested(slave_config) == 0) {
+			/* Write ready, ACK next byte */
+			I2C_SET_CONTROL_REG(i2c_base, I2C_CTL0_SI_Msk | I2C_CTL0_AA_Msk);
+		} else {
+			/* Write not ready, NACK next byte */
+			I2C_SET_CONTROL_REG(i2c_base, I2C_CTL0_SI_Msk);
+		}
+		break;
+	case S_REPEAT_START_STOP:  /* Slave Transmit/Receive Repeat Start or Stop */
+		/* Go slave end */
+		data->slave_xfer.slave_addressed = false;
+		slave_callbacks->stop(slave_config);
+		I2C_SET_CONTROL_REG(i2c_base, I2C_CTL0_SI_Msk | I2C_CTL0_AA_Msk);
+		break;
+#endif  /* CONFIG_I2C_TARGET */
+
+	case BUS_RELEASED:  /* Bus Released */
+		/* Ignore the interrupt raised by BUS_RELEASED. */
+		break;
+	default:
+		__ASSERT(false, "Uncaught I2C FSM state");
+		m_numaker_i2c_master_xfer_end(dev, true);
+	}
+}
+
+static int i2c_numaker_init(const struct device *dev)
+{
+	const struct i2c_numaker_config *config = dev->config;
+	struct i2c_numaker_data *data = dev->data;
+	int err = 0;
+	struct numaker_scc_subsys scc_subsys;
+
+	/* Validate this module's reset object */
+	if (!device_is_ready(config->reset.dev)) {
+		LOG_ERR("reset controller not ready");
+		return -ENODEV;
+	}
+
+	/* Clean mutable context */
+	memset(data, 0x00, sizeof(*data));
+
+	k_sem_init(&data->lock, 1, 1);
+	k_sem_init(&data->master_xfer.xfer_sync, 0, 1);
+
+	SYS_UnlockReg();
+
+	memset(&scc_subsys, 0x00, sizeof(scc_subsys));
+	scc_subsys.subsys_id = NUMAKER_SCC_SUBSYS_ID_PCC;
+	scc_subsys.pcc.clk_modidx = config->clk_modidx;
+	scc_subsys.pcc.clk_src = config->clk_src;
+	scc_subsys.pcc.clk_div = config->clk_div;
+
+	/* Equivalent to CLK_EnableModuleClock() */
+	err = clock_control_on(config->clkctrl_dev, (clock_control_subsys_t) &scc_subsys);
+	if (err != 0) {
+		goto cleanup;
+	}
+	/* Equivalent to CLK_SetModuleClock() */
+	err = clock_control_configure(config->clkctrl_dev,
+		(clock_control_subsys_t) &scc_subsys, NULL);
+	if (err != 0) {
+		goto cleanup;
+	}
+
+	/* Configure pinmux (NuMaker's SYS MFP) */
+#ifdef CONFIG_PINCTRL
+	err = pinctrl_apply_state(config->pincfg, PINCTRL_STATE_DEFAULT);
+	if (err != 0) {
+		goto cleanup;
+	}
```

Comment:
I think you should just `select PINCTRL` and then take it for granted that it's there, that's how nrfx does and I'd say how other platforms should do it as well, no sense failing the build just so that the user have to enable the option manually when the system can select it for them

Reply:
Sorry, this is the old-tyle compatibility code for the proprietary pin-ctl. This line was omitted during the upgrade.
Now, fixed it.